File: bfs_problems/python/binary_tree_right_side_view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        # edge case if there is not root
        if root is None:
            return []
        
        queue = [(root, 0)]
        result = [root.val]
        s = set()

        while queue:
            node, level = queue.pop()
            if node.right:
                if level not in s:
                    result.append(node.right.val)
                    s.add(level)
            if node.left:
                if level not in s:
                    result.append(node.left.val)
                    s.add(level)

            if node.left:
                queue.append((node.left, level + 1))
            if node.right:
                queue.append((node.right, level + 1))

        return result

File: dfs_problems/golang/binary_tree_right_side_view.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func dfs(root *TreeNode, result []int, level int, states map[int]bool) []int {

	if root == nil {
		return result
	}

	if root.Right != nil && root.Left != nil {
		if _, ok := states[level]; !ok {
			states[level] = true
			result = append(result, root.Right.Val)
		}
		result = dfs(root.Right, result, level+1, states)
		result = dfs(root.Left, result, level+1, states)
	} else if root.Right != nil {
		if _, ok := states[level]; !ok {
			states[level] = true
			result = append(result, root.Right.Val)
		}
		result = dfs(root.Right, result, level+1, states)
	} else if root.Left != nil {
		if _, ok := states[level]; !ok {
			states[level] = true
			result = append(result, root.Left.Val)
		}
		result = dfs(root.Left, result, level+1, states)
	}
	return result
}
func rightSideView(root *TreeNode) []int {

	var result []int
	if root == nil {
		return result
	}
	result = append(result, root.Val)
	return dfs(root, result, 0, make(map[int]bool))
}